src/tracing.md

```diff
@@ -0,0 +1,190 @@
+$ RUSTDOC_LOG=info rustdoc +stage1 my-file.rs
+
+# This will only show `debug!` calls made by rustdoc directly,
+# not any `rustc*` crate.
+$ RUSTDOC_LOG=rustdoc=debug rustdoc +stage1 my-file.rs
+```
+
+## Log colors
+
+By default, rustc (and other tools, like rustdoc and Miri) will be smart about
+when to use ANSI colors in the log output. If they are outputting to a terminal,
+they will use colors, and if they are outputting to a file or being piped
+somewhere else, they will not. However, it's hard to read log output in your
+terminal unless you have a very strict filter, so you may want to pipe the
+output to a pager like `less`. But then there won't be any colors, which makes
+it hard to pick out what you're looking for!
+
+You can override whether to have colors in log output with the `RUSTC_LOG_COLOR`
+environment variable (or `RUSTDOC_LOG_COLOR` for rustdoc, or `MIRI_LOG_COLOR`
+for Miri, etc.). There are three options: `auto` (the default), `always`, and
+`never`. So, if you want to enable colors when piping to `less`, use something
+similar to this command:
+
+```bash
+# The `-R` switch tells less to print ANSI colors without escaping them.
+$ RUSTC_LOG=debug RUSTC_LOG_COLOR=always rustc +stage1 ... | less -R
+```
+
+Note that `MIRI_LOG_COLOR` will only color logs that come from Miri, not logs
+from rustc functions that Miri calls. Use `RUSTC_LOG_COLOR` to color logs from
+rustc.
+
+## How to keep or remove `debug!` and `trace!` calls from the resulting binary
+
+While calls to `error!`, `warn!` and `info!` are included in every build of the compiler,
+calls to `debug!` and `trace!` are only included in the program if
+`debug-logging=true` is turned on in config.toml (it is
+turned off by default), so if you don't see `DEBUG` logs, especially
+if you run the compiler with `RUSTC_LOG=rustc rustc some.rs` and only see
+`INFO` logs, make sure that `debug-logging=true` is turned on in your
+config.toml.
+
+## Logging etiquette and conventions
+
+Because calls to `debug!` are removed by default, in most cases, don't worry
+about adding "unnecessary" calls to `debug!` and leaving them in code you
+commit - they won't slow down the performance of what we ship, and if they
+helped you pinning down a bug, they will probably help someone else with a
+different one.
+
+It may be preferrable to use `trace!` over `debug!` for very noisy logs.
+
+A loosely followed convention is to use `#[instrument(level = "debug")]` in
+favour of `debug!("foo(...)")` at the start of a function `foo`.
+Within functions, prefer `debug!(?variable.field)` over `debug!("xyz = {:?}", variable.field)`
+and `debug!(bar = ?var.method(arg))` over `debug!("bar = {:?}", var.method(arg))`.
+
+One thing to be **careful** of is **expensive** operations in logs.
+
+If in the module `rustc::foo` you have a statement
+
+```Rust
+debug!(x = ?random_operation(tcx));
+```
+
+Then if someone runs a debug `rustc` with `RUSTC_LOG=rustc::bar`, then
+`random_operation()` will run.
```

Comment:
this seems incorrect: if the `DEBUG` level is enabled for `rustc::bar`, but not for `rustc::foo`, then `random_operation` will not be evaluated. `random_operation` will only be evaluated if `RUSTC_LOG=rustc::foo`. 

if we recorded `random_operation` in `rustc::foo::bar`, and someone turned on `RUSTC_LOG=rustc::foo`, hoping to get debug logs from `rustc::foo::baz`, _then_ `random_operation` would be evaluated.

i think cautioning against expensive, or _especially_ potentially-crashy operations in log messages is definitely important, but the current example isn't actually accurate...

Reply:
@oli-obk maybe you meant to write "if someone runs rustc with `RUSTC_LOG=rustc`?

Reply:
this is preexisting and I only skimmed it. will reread properly

Reply:
Oh, is this copied from somewhere? Is there a way we could reduce the duplication between this and the other section? It seems silly to have two separate tutorials on how logging works ...

Reply:
Oh I see, it's moved from src/compiler-debugging.md. Yeah this is fine once you fix it to be accurate then :)

Reply:
I made this obvious by adding a commit that just does the move before the rest of this PR